### PR TITLE
Update VCPU topology handling code on AArch64

### LINKED ISSUE #3733: ACPI: AARCH64: PPTT table isn't filled correctly
The issue is observed with Windows 11 IoT guest, where just passing `--cpu boot=3` isn't sufficient and would lead to an OS crash. The guest would expect the PPTT table to be present, however it would be only generated if an explicit topology argument has been passed. See the current implementation at https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/arch/src/aarch64/fdt.rs#L222 .

According to the ARM Base Requirements 1.0 document:

```
8.3.1 Mandatory ACPI tables
...................................
8.3.1.10 PPTT
• Processor Properties Topology Table (PPTT), ACPI § 5.2.29.
o This table describes the topological structure of processors that are controlled by the OSPM and 
their shared resources.
```
Thus, this table is required for the compliance with this specification and its absence is to be seen as a bug. A meaningful topology might need to be generated, even if none has been passed explicitly, in order PPTT table to be filled correctly. Also, compared to x86_64, no explicit CPU topology argument should be required.

/cc @liuw @michael2012z @MrXinWang @jongwu 

Thanks

### PR DESCRIPTION
**Pull request by michael2012z**
Fixes #3733.

Update VCPU topology handling code:
- ACPI: When VCPU topology is not specified, a bland ACPI PPTT table was built. Now we fill the PPTT with default topology setting.
- FDT: Simplify the source code.

### CODE REVIEW

**Review by MrXinWang — APPROVED** [review]

**Comment by MrXinWang** [comment]
@michael2012z FYI I‘ve edited your PR description to fix the broken link to the issue.

File: arch/src/aarch64/fdt.rs [hunk 1]
```diff
@@ -192,32 +189,9 @@ fn create_cpu_nodes(
         fdt.end_node(cpu_node)?;
     }
 
-    // If there is a valid cpu topology config, create the cpu-map node.
-    if (threads_per_core > 0)
```
Comment (michael2012z) [inline arch/src/aarch64/fdt.rs:196]: Some explanation of the simplification:

We don't border checking if they are zero. This is done in parameter parsing.

File: arch/src/aarch64/fdt.rs [hunk 2]
```diff
@@ -192,32 +189,9 @@ fn create_cpu_nodes(
...
-        && (cores_per_package > 0)
-        && (packages > 0)
-        && (num_cpus as u8 == threads_per_core * cores_per_package * packages)
```
Comment (michael2012z) [inline arch/src/aarch64/fdt.rs:199]: This is unnecessary too. For FDT, VCPU hotplug is impossible, booting VCPU number is identical as max VCPU number. While there is check between the value of `threads_per_core * cores_per_package * package` and max VCPU number. So this is always guaranteed.

**Comment by weltling** [comment]
Thanks for the quick patch. I just went testing some option constellations. There is definitely an improvement, however some options still cause a crash.

ok:
max=2,topology=1:1:2:1
boot=2
max=2
boot=2,max=2

nok:
boot=2,max=2,topology=1:1:2:1
boot=2,topology=1:1:2:1

It looks like whenever `boot` is used together with `topology`, some invalid combinations might happen. Especially what we have on https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/cpu.md#topology as `boot=2,topology=1:1:2:1` still does crash the guest.

After a bit closer observation under gdb, it looks like `dies_per_package` part would be ignored, so then in the examples above `packages` would always arrive as `1` . Another observation also - when `boot` is omitted, it would default to `1`.  If, as told in the other comment, the hotplug won't be supported, `boot` should equal `max` then? I'd guess some validation would still be necessary, when the custom topology argument is passed. 

If it would make sense me to share some suitable image for local checks, please let me know.

Thanks

**Comment by michael2012z** [comment]
Looked into some relevant docs about the CPU topology, `dies` seems x86 specific. I guess that's why `dies_per_package` was simply ignored in AArch64 topology implementation of CH. But it definitely causes an error in calculating `max_vcpus` if `dies_per_package` is not `1`.

@weltling Can you stick `dies_per_package` to `1` in your test and only try other fields?

We will look for some improvements on handling `dies_per_package`.

**Comment by weltling** [comment]
Good catch, @michael2012z. To note also, i previously had this as a workaround using exactly `topology=1:1:1:3` which looked like a bug, that's why i didn't mention it in the initial issue description. Tried now possibliities with topology `1:1:1:3`, `1:3:1:1` and `3:1:1:1` while having `boot=3` - there's no crash and the guest would see 3 CPUs. 

So now that we know `dies_per_package` doesn't fit into the `AARCH64` reality, it's only about figuring out the best way to handle the erroneous situation it might cause.

Thanks

**Review by sboeuf — APPROVED** [review]

File: vmm/src/config.rs [hunk 3]
```diff
@@ -2188,6 +2193,14 @@ impl VmConfig {
                 return Err(ValidationError::CpuTopologyZeroPart);
             }
 
+            // The setting of dies doesen't apply on AArch64.
+            // Only '1' value is accepted, so its impact on the vcpu topology
+            // setting can be ignored.
+            #[cfg(target_arch = "aarch64")]
+            if t.dies_per_package != 1 {
```
Comment (michael2012z) [inline vmm/src/config.rs:2200]: This is the easiest way I figured out to get rid of `dies` setting on AArch64.

Alternative solution could be removing this field on AArch64 and renaming some other fields. But I think it's better to keep the parameters format aligned between architectures as much as possible.

What do you think? @sboeuf @rbradford

Comment (rbradford) [reply vmm/src/config.rs:2200]: lgtm.

**Comment by michael2012z** [comment]
Thank @weltling for testing the patches. And thanks to everybody for reviewing.

**Pull request merged**